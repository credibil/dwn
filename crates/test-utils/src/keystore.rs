use anyhow::{anyhow, Result};
use base64ct::{Base64UrlUnpadded, Encoding};
use ed25519_dalek::{SecretKey, Signer, SigningKey};
use vercre_infosec::{Algorithm, Curve, KeyType, PublicKeyJwk};

pub const OWNER_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
pub const OWNER_VERIFY_KEY: &str = "z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
pub const OWNER_SECRET_KEY: &str = "8rmFFiUcTjjrL5mgBzWykaH39D64VD0mbDHwILvsu30";

const ED25519_CODEC: [u8; 2] = [0xed, 0x01];

#[derive(Default, Clone, Debug)]
pub struct Keystore;

impl Keystore {
    pub fn try_sign(msg: &[u8]) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(OWNER_SECRET_KEY)?;
        let secret_key: SecretKey =
            decoded.try_into().map_err(|_| anyhow!("Invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.sign(msg).to_bytes().to_vec())
    }

    pub fn public_key() -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(OWNER_SECRET_KEY)?;
        let secret_key: SecretKey =
            decoded.try_into().map_err(|_| anyhow!("Invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.verifying_key().as_bytes().to_vec())
    }

    pub fn algorithm() -> Algorithm {
        Algorithm::EdDSA
    }

    pub async fn verification_method() -> Result<String> {
        Ok(format!("{OWNER_DID}#{OWNER_VERIFY_KEY}"))
    }

    pub fn public_jwk() -> PublicKeyJwk {
        let (_, key_bytes) = multibase::decode(OWNER_VERIFY_KEY).expect("should decode");
        if key_bytes.len() - 2 != 32 {
            panic!("key is not 32 bytes long");
        }
        if key_bytes[0..2] != ED25519_CODEC {
            panic!("not Ed25519");
        }

        PublicKeyJwk {
            kty: KeyType::Okp,
            crv: Curve::Ed25519,
            x: Base64UrlUnpadded::encode_string(&key_bytes[2..]),
            ..PublicKeyJwk::default()
        }
    }
}
