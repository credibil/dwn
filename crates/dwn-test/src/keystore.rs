//! # Keystore

use std::collections::HashMap;

use anyhow::{anyhow, Result};
use base64ct::{Base64UrlUnpadded, Encoding};
use ed25519_dalek::{SecretKey, Signer as _, SigningKey};
use vercre_dwn::provider::{KeyStore, Keyring};
use vercre_infosec::{Algorithm, Cipher, Signer};

use crate::store::ProviderImpl;

pub const ALICE_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
const ALICE_VERIFY_KEY: &str = "z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
const ALICE_SECRET_KEY: &str = "8rmFFiUcTjjrL5mgBzWykaH39D64VD0mbDHwILvsu30";

pub const BOB_DID: &str = "did:key:z6MkqWGVUwMwt4ahxESTVg1gjvxZ4w4KkXomksSMdCB3eHeD";
const BOB_VERIFY_KEY: &str = "z6MkqWGVUwMwt4ahxESTVg1gjvxZ4w4KkXomksSMdCB3eHeD";
const BOB_SECRET_KEY: &str = "n8Rcm64tLob0nveDUuXzP-CnLmn3V11vRqk6E3FuKCo";

pub const APP_DID: &str = "did:key:z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const APP_VERIFY_KEY: &str = "z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const APP_SECRET_KEY: &str = "fAe8yt4xBaDpyuPKY9_1NBxmiFMCfVnnryMXD-oLyVk";

// const ED25519_CODEC: [u8; 2] = [0xed, 0x01];

#[derive(Default, Clone, Debug)]
pub struct KeystoreImpl {
    keyrings: HashMap<String, KeyringImpl>,
}

#[derive(Default, Clone, Debug)]
pub struct KeyringImpl {
    pub did: String,
    pub verify_key: String,
    pub secret_key: String,
}

impl KeystoreImpl {
    // Populate the keystore with test keyrings
    pub fn new() -> Self {
        let mut keyrings = HashMap::new();

        let alice_keyring = KeyringImpl {
            did: ALICE_DID.to_string(),
            verify_key: ALICE_VERIFY_KEY.to_string(),
            secret_key: ALICE_SECRET_KEY.to_string(),
        };
        let bob_keyring = KeyringImpl {
            did: BOB_DID.to_string(),
            verify_key: BOB_VERIFY_KEY.to_string(),
            secret_key: BOB_SECRET_KEY.to_string(),
        };
        let app_keyring = KeyringImpl {
            did: APP_DID.to_string(),
            verify_key: APP_VERIFY_KEY.to_string(),
            secret_key: APP_SECRET_KEY.to_string(),
        };

        keyrings.insert(ALICE_DID.to_string(), alice_keyring);
        keyrings.insert(BOB_DID.to_string(), bob_keyring);
        keyrings.insert(APP_DID.to_string(), app_keyring);

        Self { keyrings }
    }
}

impl KeyStore for ProviderImpl {
    fn keyring(&self, identifier: &str) -> Result<impl Keyring> {
        self.keystore.keyrings.get(identifier).cloned().ok_or_else(|| anyhow!("Keyring not found"))
    }
}

impl Keyring for KeyringImpl {}

impl Signer for KeyringImpl {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: SecretKey =
            decoded.try_into().map_err(|_| anyhow!("Invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.sign(msg).to_bytes().to_vec())
    }

    async fn public_key(&self) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: SecretKey =
            decoded.try_into().map_err(|_| anyhow!("Invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.verifying_key().as_bytes().to_vec())
    }

    fn algorithm(&self) -> Algorithm {
        Algorithm::EdDSA
    }

    async fn verification_method(&self) -> Result<String> {
        Ok(format!("{}#{}", self.did, self.verify_key))
    }
}

impl Cipher for KeyringImpl {
    async fn encrypt(&self, _plaintext: &[u8], _recipient_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }

    fn ephemeral_public_key(&self) -> Vec<u8> {
        todo!()
    }

    async fn decrypt(&self, _ciphertext: &[u8], _sender_public_key: &[u8]) -> Result<Vec<u8>> {
        todo!()
    }
}

// #[derive(Default, Clone, Debug)]
// pub struct Keystore;

// impl Keystore {
//     pub fn try_sign(&self, _msg: &[u8]) -> Result<Vec<u8>> {
//         unimplemented!()
//     }

//     pub fn public_key(&self) -> Result<Vec<u8>> {
//         unimplemented!()
//     }

//     pub fn algorithm(&self) -> Algorithm {
//         Algorithm::EdDSA
//     }

//     pub async fn verification_method(&self) -> Result<String> {
//         unimplemented!()
//     }

//     pub fn public_jwk(&self) -> PublicKeyJwk {
//         unimplemented!()

//         // let (_, key_bytes) = multibase::decode(OWNER_VERIFY_KEY).expect("should decode");
//         // if key_bytes.len() - 2 != 32 {
//         //     panic!("key is not 32 bytes long");
//         // }
//         // if key_bytes[0..2] != ED25519_CODEC {
//         //     panic!("not Ed25519");
//         // }

//         // PublicKeyJwk {
//         //     kty: KeyType::Okp,
//         //     crv: Curve::Ed25519,
//         //     x: Base64UrlUnpadded::encode_string(&key_bytes[2..]),
//         //     ..PublicKeyJwk::default()
//         // }
//     }
// }
