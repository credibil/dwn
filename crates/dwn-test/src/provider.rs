#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
pub mod event_log;
mod event_stream;
pub mod key_store;
pub mod message_store;
pub mod task_store;

use anyhow::{Result, anyhow};
use blockstore::InMemoryBlockstore;
use surrealdb::Surreal;
use surrealdb::engine::local::{Db, Mem};
use vercre_dwn_server::provider::{DidResolver, Document, Provider};

use self::key_store::{ALICE_DID, KeyStoreImpl};

const NAMESPACE: &str = "integration-test";

#[derive(Clone)]
pub struct ProviderImpl {
    db: Surreal<Db>,
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
    pub keystore: KeyStoreImpl,
}

impl Provider for ProviderImpl {}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        let db = Surreal::new::<Mem>(()).await?;
        db.use_ns(NAMESPACE).use_db(ALICE_DID).await?;

        Ok(Self {
            db,
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
            keystore: KeyStoreImpl::new(),
        })
    }
}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        if url == ALICE_DID {
            return serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        } else {
            return serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
                .map_err(|e| anyhow!(format!("issue deserializing document: {e}")));
        }
    }
}
