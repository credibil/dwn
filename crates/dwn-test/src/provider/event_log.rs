use std::collections::BTreeMap;

use anyhow::Result;
use async_trait::async_trait;
use serde_json::Value;
use vercre_dwn::event::Event;
use vercre_dwn::provider::EventLog;
use vercre_dwn::store::{Cursor, Query};
use vercre_serialize::QuerySerializer;

use super::ProviderImpl;
use crate::provider::NAMESPACE;

const TABLE: &str = "event_log";

#[async_trait]
impl EventLog for ProviderImpl {
    async fn append(&self, owner: &str, event: &Event) -> Result<()> {
        self.db.use_ns(NAMESPACE).use_db(owner).await?;
        let _: Option<BTreeMap<String, Value>> =
            self.db.create((TABLE, &event.cid()?)).content(event).await?;
        Ok(())
    }

    async fn events(&self, owner: &str, cursor: Option<Cursor>) -> Result<(Vec<Event>, Cursor)> {
        todo!()
    }

    async fn query(&self, owner: &str, query: &Query) -> Result<(Vec<Event>, Cursor)> {
        let sql = query.serialize();
        // println!("{}", sql);
        let mut response = self.db.query(&sql).bind(("table", TABLE)).await?;
        let events: Vec<Event> = response.take(0)?;
        Ok((events, Cursor::default()))

        // TODO: sort and paginate
    }

    async fn delete(&self, owner: &str, message_cid: &str) -> Result<()> {
        self.db.use_ns(NAMESPACE).use_db(owner).await?;
        let _: Option<BTreeMap<String, Value>> = self.db.delete((TABLE, message_cid)).await?;
        Ok(())
    }

    async fn purge(&self) -> Result<()> {
        todo!()
    }
}
