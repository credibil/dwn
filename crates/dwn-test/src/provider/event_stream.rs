use anyhow::Result;
use async_trait::async_trait;
use futures::stream::StreamExt;
use vercre_dwn::event::{Event, Subscriber};
use vercre_dwn::provider::EventStream;

use crate::provider::ProviderImpl;

const SUBJECT: &str = "events";

#[async_trait]
impl EventStream for ProviderImpl {
    /// Subscribe to a owner's event stream.
    async fn subscribe(&self, owner: &str) -> Result<Subscriber> {
        let subscriber = self.nats_client.subscribe(format!("{SUBJECT}.{owner}")).await?;
        let mapped = subscriber.map(|m| serde_json::from_slice::<Event>(&m.payload).unwrap());
        Ok(Subscriber::new(mapped))
    }

    /// Emits an event to a owner's event stream.
    async fn emit(&self, owner: &str, event: &Event) -> Result<()> {
        let bytes = serde_json::to_vec(event)?;
        self.nats_client.publish(format!("{SUBJECT}.{owner}"), bytes.into()).await?;
        Ok(())
    }
}
