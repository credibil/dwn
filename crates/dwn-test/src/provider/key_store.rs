//! # Keystore

use std::collections::HashMap;

use anyhow::{Result, anyhow};
use base64ct::{Base64UrlUnpadded, Encoding};
use ed25519_dalek::{Signer as _, SigningKey};
use vercre_dwn::provider::{KeyStore, Keyring};
use vercre_infosec::{Algorithm, PublicKey, Receiver, SecretKey, SharedSecret, Signer};

// use x25519_dalek::{PublicKey, StaticSecret};
use crate::provider::ProviderImpl;

pub const ALICE_DID: &str = "did:key:z6Mkj8Jr1rg3YjVWWhg7ahEYJibqhjBgZt1pDCbT4Lv7D4HX";
const ALICE_SECRET_KEY: &str = "8rmFFiUcTjjrL5mgBzWykaH39D64VD0mbDHwILvsu30";

pub const BOB_DID: &str = "did:key:z6MkqWGVUwMwt4ahxESTVg1gjvxZ4w4KkXomksSMdCB3eHeD";
const BOB_SECRET_KEY: &str = "n8Rcm64tLob0nveDUuXzP-CnLmn3V11vRqk6E3FuKCo";

pub const CAROL_DID: &str = "did:key:z6MkuY2MjELw3xQExptJtVkuW5YSjfeQYrVZv41RMYrwZmYd";
const CAROL_SECRET_KEY: &str = "V0YsmES1Tc8-sozoyYeBKemcrUaOLq_IceWbjWwmbMo";

pub const APP_DID: &str = "did:key:z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const APP_SECRET_KEY: &str = "fAe8yt4xBaDpyuPKY9_1NBxmiFMCfVnnryMXD-oLyVk";

pub const INVALID_DID: &str = "did:key:z6Mkj85hWKz3rvxVt6gL54rCsEMia8ZRXMTmxaUv4yLDSnTA";
const INVALID_SECRET_KEY: &str = "n8Rcm64tLob0nveDUuXzP-CnLmn3V11vRqk6E3FuKCo";

#[derive(Default, Clone, Debug)]
pub struct KeyStoreImpl {
    keyrings: HashMap<String, KeyringImpl>,
}

#[derive(Default, Clone, Debug)]
pub struct KeyringImpl {
    pub did: String,
    pub secret_key: String,
}

impl KeyStoreImpl {
    // Populate the keystore with test keyrings
    pub fn new() -> Self {
        let mut keyrings = HashMap::new();

        let alice_keyring = KeyringImpl {
            did: ALICE_DID.to_string(),
            secret_key: ALICE_SECRET_KEY.to_string(),
        };
        let bob_keyring = KeyringImpl {
            did: BOB_DID.to_string(),
            secret_key: BOB_SECRET_KEY.to_string(),
        };
        let carol_keyring = KeyringImpl {
            did: CAROL_DID.to_string(),
            secret_key: CAROL_SECRET_KEY.to_string(),
        };
        let app_keyring = KeyringImpl {
            did: APP_DID.to_string(),
            secret_key: APP_SECRET_KEY.to_string(),
        };
        let invalid_keyring = KeyringImpl {
            did: INVALID_DID.to_string(),
            secret_key: INVALID_SECRET_KEY.to_string(),
        };

        keyrings.insert(ALICE_DID.to_string(), alice_keyring);
        keyrings.insert(BOB_DID.to_string(), bob_keyring);
        keyrings.insert(CAROL_DID.to_string(), carol_keyring);
        keyrings.insert(APP_DID.to_string(), app_keyring);
        keyrings.insert(INVALID_DID.to_string(), invalid_keyring);

        Self { keyrings }
    }
}

impl KeyStore for ProviderImpl {
    fn keyring(&self, identifier: &str) -> Result<impl Keyring> {
        self.keystore.keyrings.get(identifier).cloned().ok_or_else(|| anyhow!("keyring not found"))
    }
}

impl Keyring for KeyringImpl {}

impl Signer for KeyringImpl {
    async fn try_sign(&self, msg: &[u8]) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: ed25519_dalek::SecretKey =
            decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.sign(msg).to_bytes().to_vec())
    }

    async fn public_key(&self) -> Result<Vec<u8>> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let secret_key: ed25519_dalek::SecretKey =
            decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;
        let signing_key: SigningKey = SigningKey::from_bytes(&secret_key);

        Ok(signing_key.verifying_key().as_bytes().to_vec())
    }

    fn algorithm(&self) -> Algorithm {
        Algorithm::EdDSA
    }

    async fn verification_method(&self) -> Result<String> {
        let verify_key = self.did.strip_prefix("did:key:").unwrap_or_default();
        Ok(format!("{}#{}", self.did, verify_key))
    }
}

impl Receiver for KeyringImpl {
    // // FIXME: wrap return in Result<Vec<u8>>
    // fn public_key(&self) -> Vec<u8> {
    //     let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key).unwrap();
    //     let bytes: [u8; 32] =
    //         decoded.try_into().map_err(|_| anyhow!("Invalid secret key")).unwrap();

    //     // derive x25519 key agreement public key
    //     PublicKey::from(bytes).as_bytes().to_vec()
    // }

    fn key_id(&self) -> String {
        self.did.clone()
    }

    async fn shared_secret(&self, sender_public: PublicKey) -> Result<SharedSecret> {
        let decoded = Base64UrlUnpadded::decode_vec(&self.secret_key)?;
        let bytes: [u8; 32] = decoded.try_into().map_err(|_| anyhow!("invalid secret key"))?;

        let secret_key = SecretKey::from(bytes);
        secret_key.shared_secret(sender_public)
    }
}
