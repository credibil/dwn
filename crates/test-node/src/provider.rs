#![allow(missing_docs)]
#![allow(unused_variables)]

//! # Provider
//!
//! Implementation of the `Provider` trait for testing and examples.

pub mod block_store;
mod event_stream;
pub mod key_store;

use anyhow::Result;
use blockstore::InMemoryBlockstore;
use dwn_node::provider::{
    DataStore, DidResolver, Document, EventLog, MessageStore, Provider, TaskStore,
};

#[derive(Clone)]
pub struct ProviderImpl {
    blockstore: InMemoryBlockstore<64>,
    pub nats_client: async_nats::Client,
}

impl ProviderImpl {
    pub async fn new() -> Result<Self> {
        Ok(Self {
            blockstore: InMemoryBlockstore::<64>::new(),
            nats_client: async_nats::connect("demo.nats.io").await?,
        })
    }
}

impl Provider for ProviderImpl {}
impl MessageStore for ProviderImpl {}
impl DataStore for ProviderImpl {}
impl EventLog for ProviderImpl {}
impl TaskStore for ProviderImpl {}

impl DidResolver for ProviderImpl {
    async fn resolve(&self, url: &str) -> Result<Document> {
        // let resolved =
        //     vercre_did::resolve(url, None, self.clone()).await.map_err(|e| anyhow!(e))?;
        // Ok(resolved.document.unwrap())

        // if url == &alice.did {
        //     serde_json::from_slice(include_bytes!("./provider/data/alice_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // } else {
        //     serde_json::from_slice(include_bytes!("./provider/data/bob_did.json"))
        //         .map_err(|e| anyhow!(format!("issue deserializing document: {e}")))
        // }

        println!("resolve: {}", url);

        Ok(Document::default())
    }
}
