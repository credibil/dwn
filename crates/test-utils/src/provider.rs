//! # Mock Provider

use std::sync::Arc;

use anyhow::Result;
use blockstore::{Blockstore as _, InMemoryBlockstore};
use cid::Cid;
use credibil_core::api::Client;
use credibil_dwn::event::{Event, Subscriber};
use credibil_dwn::provider::{BlockStore, EventStream, Resolver};
use credibil_proof::DocumentRequest;
use futures::stream::StreamExt;
use multihash_codetable::MultihashDigest;
use serde::{Deserialize, Serialize};

use crate::store::Store;

const RAW: u64 = 0x55;
const SUBJECT: &str = "events";
#[derive(Clone)]
pub struct Provider {
    blockstore: Arc<InMemoryBlockstore<64>>,
    pub nats_client: Arc<async_nats::Client>,
}

impl Provider {
    pub async fn new() -> Self {
        Self {
            blockstore: Arc::new(InMemoryBlockstore::<64>::new()),
            nats_client: Arc::new(
                async_nats::connect("demo.nats.io").await.expect("should connect"),
            ),
        }
    }
}

impl BlockStore for Provider {
    async fn put(&self, owner: &str, partition: &str, key: &str, block: &[u8]) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.put_keyed(&cid, block).await.map_err(Into::into)
    }

    async fn get(&self, owner: &str, partition: &str, key: &str) -> Result<Option<Vec<u8>>> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        let Some(bytes) = self.blockstore.get(&cid).await? else {
            return Ok(None);
        };
        Ok(Some(bytes))
    }

    async fn delete(&self, owner: &str, partition: &str, key: &str) -> Result<()> {
        let cid = Identitifier::new(owner, partition, key).to_cid()?;
        self.blockstore.remove(&cid).await?;
        Ok(())
    }

    async fn purge(&self, _owner: &str, _partition: &str) -> Result<()> {
        unimplemented!()
    }
}

#[derive(Serialize, Deserialize)]
struct Identitifier<'a> {
    owner: &'a str,
    partition: &'a str,
    key: &'a str,
}

impl<'a> Identitifier<'a> {
    fn new(owner: &'a str, partition: &'a str, key: &'a str) -> Self {
        Self {
            owner,
            partition,
            key,
        }
    }

    fn to_cid(&self) -> anyhow::Result<Cid> {
        let mut buf = Vec::new();
        ciborium::into_writer(self, &mut buf)?;
        let hash = multihash_codetable::Code::Sha2_256.digest(&buf);
        Ok(Cid::new_v1(RAW, hash))
    }
}

impl Resolver for Provider {
    async fn resolve(&self, url: &str) -> Result<Vec<u8>> {
        let request = DocumentRequest { url: url.to_string() };
        let document = Client::new(Store)
            .request(request)
            .owner("owner")
            .execute()
            .await
            .map(|r| r.0.clone())?;
        serde_json::to_vec(&document).map_err(|e| e.into())
    }
}

impl EventStream for Provider {
    /// Subscribe to a owner's event stream.
    async fn subscribe(&self, owner: &str) -> Result<Subscriber> {
        let subscriber = self.nats_client.subscribe(format!("{SUBJECT}.{owner}")).await?;
        let mapped = subscriber.map(|m| serde_json::from_slice::<Event>(&m.payload).unwrap());
        Ok(Subscriber::new(mapped))
    }

    /// Emits an event to a owner's event stream.
    async fn emit(&self, owner: &str, event: &Event) -> Result<()> {
        let bytes = serde_json::to_vec(event)?;
        self.nats_client.publish(format!("{SUBJECT}.{owner}"), bytes.into()).await?;
        Ok(())
    }
}
