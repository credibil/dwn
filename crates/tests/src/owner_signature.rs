//! Owner Signature
//!
//! This test demonstrates how a web node owner can delegate permission to
//! another entity to perform an action on their behalf. In this case, Alice
//! grants Bob the ability to configure a protocol on her behalf.

use std::io::Read;
use std::sync::LazyLock;

use dwn_node::endpoint;
use dwn_node::interfaces::records::{Data, ReadBuilder, RecordsFilter, WriteBuilder};
use http::StatusCode;
use serde_json::{Value, json};
use test_node::keystore::{self, Keyring};
use test_node::provider::ProviderImpl;

static ALICE: LazyLock<Keyring> = LazyLock::new(|| keystore::new_keyring());
static BOB: LazyLock<Keyring> = LazyLock::new(|| keystore::new_keyring());

// Use owner signature for authorization when it is provided.
#[tokio::test]
async fn flat_space() {
    let provider = ProviderImpl::new().await.expect("should create provider");

    // --------------------------------------------------
    // Bob writes a message to his web node
    // --------------------------------------------------
    let bob_data = serde_json::to_vec(&json!({
        "message": "test record write",
    }))
    .expect("should serialize");

    let bob_msg = WriteBuilder::new()
        .data(Data::from(bob_data))
        .published(true)
        .sign(&*BOB)
        .build()
        .await
        .expect("should create write");

    let reply = endpoint::handle(&BOB.did, bob_msg.clone(), &provider).await.expect("should write");
    assert_eq!(reply.status.code, StatusCode::ACCEPTED);

    // --------------------------------------------------
    // Alice fetches the message from Bob's web node
    // --------------------------------------------------
    let filter = RecordsFilter::new().record_id(&bob_msg.record_id);
    let alice_read =
        ReadBuilder::new().filter(filter).sign(&*ALICE).build().await.expect("should create write");

    let reply =
        endpoint::handle(&BOB.did, alice_read.clone(), &provider).await.expect("should read");
    assert_eq!(reply.status.code, StatusCode::OK);

    let read_reply = reply.body.expect("should be records read");
    let mut read_bob_msg = read_reply.entry.records_write.expect("should have records write entry");
    assert_eq!(read_bob_msg.record_id, bob_msg.record_id);

    // --------------------------------------------------
    // Alice augments Bob's message and saves to her web node
    // --------------------------------------------------
    let alice_data = read_reply.entry.data.expect("should have data");
    read_bob_msg.sign_as_owner(&*ALICE).await.expect("should sign as owner");
    read_bob_msg.with_stream(alice_data);

    let reply = endpoint::handle(&ALICE.did, read_bob_msg, &provider).await.expect("should write");
    assert_eq!(reply.status.code, StatusCode::ACCEPTED);

    // --------------------------------------------------
    // Bob's message can be read from Alice's web node
    // --------------------------------------------------
    let reply = endpoint::handle(&BOB.did, alice_read, &provider).await.expect("should read");
    assert_eq!(reply.status.code, StatusCode::OK);

    let read_reply = reply.body.expect("should be records read");
    let mut reader = read_reply.entry.data.expect("should have data");
    let mut alice_data = Vec::new();
    reader.read_to_end(&mut alice_data).expect("should read to end");

    let bob_data: Value = serde_json::from_slice(&alice_data).expect("should deserialize");
    assert_eq!(json! ({ "message": "test record write" }), bob_data);
}
