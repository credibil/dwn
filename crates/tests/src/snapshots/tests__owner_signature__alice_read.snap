---
source: crates/tests/src/owner_signature.rs
assertion_line: 57
expression: reply
---
status:
  code: 200
entry:
  recordsWrite:
    descriptor:
      interface: Records
      method: Write
      messageTimestamp: "[messageTimestamp]"
      dataCid: bafkreigu2kgwt6jz332rro7iotzmkiy74mdpux744wqcqf2mxh7l2hm6jq
      dataSize: 31
      dataFormat: application/json
      dateCreated: "[dateCreated]"
      published: true
      datePublished: "[datePublished]"
    authorization:
      signature:
        payload: "[payload]"
        signatures:
          - protected: eyJhbGciOiJFZERTQSIsInR5cCI6Imp3dCIsImtpZCI6ImRpZDprZXk6ejZNa3FXR1ZVd013dDRhaHhFU1RWZzFnanZ4WjR3NEtrWG9ta3NTTWRDQjNlSGVEI3o2TWtxV0dWVXdNd3Q0YWh4RVNUVmcxZ2p2eFo0dzRLa1hvbWtzU01kQ0IzZUhlRCJ9
            signature: "[signature]"
    recordId: "[recordId]"
